### PR TITLE
Add automatic JDK installation via Jabba

### PR DESCRIPTION
**Pull request by sswistun-vl**
Previously, the VS Code extension would fail with an error message when it was unable to find a valid Java installation on the computer. Now, the extension asks the user to automatically install Java version via Jabba (https://github.com/shyiko/jabba) when there is no valid Java installation on the computer or if there's only Java 9.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -96,11 +122,14 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
...
     outputChannel.appendLine(
       `Metals will not start since Dotty is enabled for this workspace. ` +
-        `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
+      `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
     );
     return;
   }
 
+  if (!javaHome || javaHome.length == 0 || !fs.existsSync(javaHome))
+    throw new Error("Invalid Java Home path");
```
Comment (olafurpg) [inline src/extension.ts]: Should we report an error message to the user instead of throwing an exception?

File: src/jdkInstaller.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            }
+            else res();
+        });
+    }).then(() => curl(jabbaUrl, jabba, true)
+        .then(pth => pcp.exec(`${jabba} ls-remote`))
+        .then(out => outputToString(out.stdout).split("\n")
+            .filter((str, i, arr) => str.includes(javaVersion))[0].trim())
+        .then(async java => {
+            if (out) out(`Installing ${java}`);
+            await pcp.exec(`${jabba} install ${java}`);
```
Comment (olafurpg) [inline src/jdkInstaller.ts]: Do we display any progress bar to the user while this download happens? It would be great to stream stdout to the Metals `OutputChannel` and make sure that the output channel is visible during the download.

File: src/extension.ts [hunk 3]
```diff
@@ -96,11 +122,14 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
   if (dottyArtifact && fs.existsSync(dottyArtifact)) {
     outputChannel.appendLine(
       `Metals will not start since Dotty is enabled for this workspace. ` +
-        `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
+      `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
```
Comment (olafurpg) [inline src/extension.ts]: Was this change intentional? Make sure to use Prettier for formatting the files

File: src/extension.ts [hunk 4]
```diff
@@ -58,12 +58,21 @@ import { startTreeView } from "./treeview";
 import { MetalsFeatures } from "./MetalsFeatures";
 import { MetalsTreeViewReveal, MetalsTreeViews } from "./tree-view-protocol";
 import * as scalaDebugger from "./scalaDebugger";
+import * as jdki from "./jdkInstaller";
```
Comment (olafurpg) [inline src/extension.ts]: Can we try to use consistent names between the file and imported variable?
```suggestion
import * as jdkInstaller from "./jdkInstaller";
```

**Comment by olafurpg** [comment]
I took the liberty to update the PR description and title according to https://github.com/scalameta/metals-vscode/pull/161#issuecomment-550147732 Please try to use this format in the future.

File: src/extension.ts (hunk 1 above)
Comment (sswistun-vl) [reply src/extension.ts]: It gets reported to user when it's caught by catch at line 82

**Comment by sswistun-vl** [comment]
There is one issue that I've noticed - adopt JDK uses different version scheme than others - for example: adopt@1.8.0-232 and amazon-corretto@1.8.232-09.1.

File: src/jdkInstaller.ts (hunk 2 above)
Comment (sswistun-vl) [reply src/jdkInstaller.ts]: Did you meant something like what I've implemented?

File: src/extension.ts (hunk 3 above)
Comment (sswistun-vl) [reply src/extension.ts]: Fixed.

**Comment by tgodzik** [comment]
@olafurpg what do you think about the change? We also added the Java 9 use case - which should help with https://github.com/scalameta/metals/issues/1037

**Comment by tgodzik** [comment]
@sswistun-vl Let's rebase this PR

**Comment by sswistun-vl** [comment]
Rebased.

File: src/extension.ts [hunk 5]
```diff
@@ -58,29 +55,72 @@ import { startTreeView } from "./treeview";
...
+        .then(choice => {
+          const jdkinstaller = new JdkInstaller(outputChannel);
+          switch (choice) {
+            case openSettingsAction: {
+              commands.executeCommand("workbench.action.openSettings");
+              break;
+            }
+            case jdk8Action: {
+              window.showInformationMessage(
+                `Installing ${jdk8}, please wait...`
```
[the remark this answers is not in the document]
Comment (sswistun-vl) [reply src/extension.ts]: It's handled by Jabba.

File: src/jdkInstaller.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  isWindows(): boolean {
+    return process.platform === "win32";
+  }
+
+  jabbaName(): string {
+    if (this.isWindows()) return "jabba.exe";
+    else return "jabba";
+  }
+
+  installJava(javaVersion: string, jabbaVersion: string) {
```
[the remark this answers is not in the document]
Comment (sswistun-vl) [reply src/jdkInstaller.ts]: Done.

File: package.json [hunk 7]
```diff
@@ -98,6 +98,11 @@
           "type": "string",
           "markdownDescription": "Optional path to the Java home directory. Requires reloading the window.\n\nDefaults to the most recent Java 8 version computed by the `locate-java-home` npm package."
         },
+        "metals.jabbaVersion": {
```
[the remark this answers is not in the document]
Comment (sswistun-vl) [reply package.json]: Removed.

File: src/extension.ts [hunk 8]
```diff
@@ -58,29 +55,72 @@ import { startTreeView } from "./treeview";
...
   checkServerVersion();
 
   getJavaHome()
     .then(javaHome => fetchAndLaunchMetals(context, javaHome))
     .catch(err => {
-      const message =
-        "Unable to find a Java 8 or Java 11 installation on this computer. " +
-        "To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory";
+      var message = "";
+      if (err.message === "Java9") {
```
[the remark this answers is not in the document]
Comment (sswistun-vl) [reply src/extension.ts]: I've extracted "Java9" to variable. This check is used to determine if problem is with Java version or lack of it.

File: src/extension.ts [hunk 9]
```diff
@@ -101,11 +141,13 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
     return;
   }
 
+  if (!javaHome || javaHome.trim().length == 0 || !fs.existsSync(javaHome))
+    throw new Error("Invalid JavaHome");
```
[the remark this answers is not in the document]
Comment (sswistun-vl) [reply src/extension.ts]: This exception is handled at line 119. User gets error message when it's thrown.

File: src/extension.ts [hunk 10]
```diff
@@ -79,23 +79,68 @@ let decorationType: TextEditorDecorationType = window.createTextEditorDecoration
...
   checkServerVersion();
 
   getJavaHome()
     .then(javaHome => fetchAndLaunchMetals(context, javaHome))
     .catch(err => {
-      const message =
-        "Unable to find a Java 8 or Java 11 installation on this computer. " +
-        "To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory";
+      var message = "";
+      if (err.message === java9) {
```
Comment (olafurpg) [inline src/extension.ts]: nit: can we extract this if/else chain into a method so we can do `const message = getMessage(err) + "To fix this problem ..."`

File: src/extension.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (olafurpg) [reply src/extension.ts]: Can we use something more robust than the stringly typed logic `err.message.includes("JavaHome")`?

File: src/extension.ts [hunk 11]
```diff
@@ -228,6 +275,11 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
   );
 }
 
+function updateJavaConfig(javaHome: string, global: boolean = true) {
+  workspace.getConfiguration("metals").update("javaHome", javaHome, global);
+  commands.executeCommand("workbench.action.reloadWindow");
```
Comment (olafurpg) [inline src/extension.ts]: Let's not reload automatically, Metals will already prompt the user to reload the window when settings change. Some users may have state in their terminal that they don't want to loose

File: src/extension.ts [hunk 12]
```diff
@@ -841,22 +893,18 @@ function migrateStringSettingToArray(id: string): void {
...
 
   if (typeof setting.globalValue === "string") {
-    workspace
-      .getConfiguration("metals")
-      .update(
-        id,
-        setting.globalValue.split(" ").filter(e => e.length > 0),
-        ConfigurationTarget.Global
-      );
+    workspace.getConfiguration("metals").update(
```
Comment (olafurpg) [inline src/extension.ts]: Is this formatting change caused by a Prettier version mis-match? @gabro is there a way to enforce a stable version of Prettier in CI?

Comment (gabro) [reply src/extension.ts]: Yes, it's sufficient to `yarn add --dev prettier` and then run `yarn prettier --check` on the CI.

File: src/extension.ts (hunk 11 above)
Comment (gabro) [reply src/extension.ts]: please note that there's a generalized way to do this in response to configuration changes. See https://github.com/scalameta/metals-vscode/blob/136ef625062b690e9ac22a9464c8ac8d281a7d1a/src/extension.ts#L729-L754

Comment (gabro) [reply src/extension.ts]: (actually, `javaHome` is already in there, so the user should be prompted automatically already after the `update`)

File: src/extension.ts [hunk 13]
```diff
@@ -79,23 +79,68 @@ let decorationType: TextEditorDecorationType = window.createTextEditorDecoration
   }
 );
 
+const jdk8 = "1.8.232";
+const jdk8Action = "JDK8";
+
+const jdk11 = "1.11";
+const jdk11Action = "JDK11";
+
+const default_jabba_ver = "0.11.2";
```
Comment (gabro) [inline src/extension.ts]: nit: please use camel case for identifier names.

```suggestion
const defaultJabbaVer = "0.11.2";
```

File: src/extension.ts (hunk 10 above)
Comment (gabro) [reply src/extension.ts]: I agree, this would also allow avoiding mutation.

By the way, @sswistun-vl even if we need a mutable variable, I recommend using `let` instead of `var`, since `let` has way more intuitive and sensible scoping rules (`let` is block-scoped whereas `var` is function-scoped)

File: src/extension.ts [hunk 14]
```diff
@@ -79,23 +79,68 @@ let decorationType: TextEditorDecorationType = window.createTextEditorDecoration
...
+
+const config = workspace.getConfiguration("metals");
+
+const java9 = "Java9";
```
Comment (gabro) [inline src/extension.ts]: Suggestion: I would name this constant something more telling, such as `unsupportedJavaVersion9` or something like that.

Also, I would use the same constant where you raise the error, so that if you change it in the future you only have one source of truth for the underlying string.

Comment (gabro) [reply src/extension.ts]: as an aside, did I mention that I really really dislike the `Promise` API because it doesn't carry the error type? Nothing related to this PR, feel free to ignore my rants :D

File: src/extension.ts (hunk 10 above)
Comment (sswistun-vl) [reply src/extension.ts]: Changed, good to know about var and let difference, I don't know too much about js/ts.

File: src/extension.ts (hunk 11 above)
Comment (sswistun-vl) [reply src/extension.ts]: I've removed auto-reload.

File: src/extension.ts (hunk 13 above)
Comment (sswistun-vl) [reply src/extension.ts]: Changed.

File: src/extension.ts (hunk 14 above)
Comment (sswistun-vl) [reply src/extension.ts]: Renamed. I understand why you dislike it, but on the other hand it makes chaining futures much clearer to write and read :D

Comment (gabro) [reply src/extension.ts]: Oh, I'm all for the chaining API, I'm just bothered by the specific implementation :P 

I may propose a rewrite using `TaskEither` from fp-ts in the future ;)

File: src/extension.ts (hunk 12 above)
Comment (tgodzik) [reply src/extension.ts]: Added an issue for this.

File: src/extension.ts [hunk 15]
```diff
@@ -87,17 +97,58 @@ export async function activate(context: ExtensionContext) {
...
+    .then(choice => {
+      const jdkinstaller = new JdkInstaller(outputChannel);
+      switch (choice) {
+        case openSettingsAction: {
           commands.executeCommand("workbench.action.openSettings");
+          break;
         }
-      });
+        case jdk8Action: {
+          window.showInformationMessage(`Installing ${jdk8}, please wait...`);
```
Comment (tgodzik) [inline src/extension.ts]: Can we use `window.withProgress` here? Tested it out and it looks really cool, the only issue I have is the window just shows up and it's hard to say what is going on, it would be nice to have:
1. `Installing ${jdk8}, please wait...` showing progress (or just 100% when it's done)
2. Then the window disappears and reload suggestion appears.

Comment (tgodzik) [reply src/extension.ts]: Ok, that's fixed and it shows progress nicely

File: package.json [hunk 16]
```diff
@@ -251,6 +252,8 @@
   },
   "dependencies": {
     "metals-languageclient": "^0.1.6",
+    "@types/request": "^2.48.3",
```
Comment (gabro) [inline package.json]: this should be a devDependency

File: src/extension.ts [hunk 17]
```diff
@@ -78,6 +78,16 @@ let decorationType: TextEditorDecorationType = window.createTextEditorDecoration
   }
 );
 
+const jdk8 = "1.8.232";
+const jdk8Action = "JDK8";
+
+const jdk11 = "1.11";
```
Comment (gabro) [inline src/extension.ts]: are you sure about this? We recently fixed a bug due to Java 11 being on version “11” rather than “1.11”

Comment (tgodzik) [reply src/extension.ts]: It's coming from https://github.com/shyiko/jabba/blob/master/index.json

Renamed it to `jdk11Jabba` to pronounce it better.

File: package.json (hunk 16 above)
Comment (tgodzik) [reply package.json]: Moved it, thanks!

File: src/extension.ts [hunk 18]
```diff
@@ -87,16 +97,53 @@ export async function activate(context: ExtensionContext) {
     .catch(err => {
       const message =
         "Unable to find a Java 8 or Java 11 installation on this computer. " +
-        "To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory";
+        "To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory " +
+        "or select version to install automatically (window will need to be reloaded afterwards)";
```
Comment (ckipp01) [inline src/extension.ts]:
```suggestion
        "or select a version to install automatically (window will need to be reloaded afterwards)";
```

File: src/extension.ts [hunk 19]
```diff
@@ -110,7 +157,15 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
...
-        `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
+      `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
+    );
+    return;
+  }
+
+  if (!javaHome || javaHome.trim().length == 0 || !fs.existsSync(javaHome)) {
+    showJabbaMessage(
+      "Incorrect 'Java Home' setting. To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory " +
+      "or select version to install automatically (window will need to be reloaded afterwards)."
```
Comment (ckipp01) [inline src/extension.ts]:
```suggestion
      "or select a version to install automatically (window will need to be reloaded afterwards)."
```

Comment (tgodzik) [reply src/extension.ts]: Thanks! Fixed both!

**Comment by tgodzik** [comment]
@gabro Are you ok with merging this?

File: src/extension.ts [hunk 20]
```diff
@@ -87,16 +97,53 @@ export async function activate(context: ExtensionContext) {
     .catch(err => {
       const message =
         "Unable to find a Java 8 or Java 11 installation on this computer. " +
-        "To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory";
+        "To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory " +
+        "or select a version to install automatically (window will need to be reloaded afterwards)";
```
Comment (gabro) [inline src/extension.ts]: nit: instead of writing it here, can we automatically prompt the user to reload the window after installation?

File: src/extension.ts [hunk 21]
```diff
@@ -87,16 +97,53 @@ export async function activate(context: ExtensionContext) {
...
+
       outputChannel.appendLine(message);
       outputChannel.appendLine(err);
-      window.showErrorMessage(message, openSettingsAction).then(choice => {
-        if (choice === openSettingsAction) {
+
+      const jabba_ver = defaultJabbaVer;
```
Comment (gabro) [inline src/extension.ts]: nit: 

```suggestion
      const jabbaVersion = defaultJabbaVer;
```

Comment (gabro) [reply src/extension.ts]: I don't see the point of this const, since it's only used in the next call. I would remove it.

File: src/extension.ts [hunk 22]
```diff
@@ -78,6 +78,16 @@ let decorationType: TextEditorDecorationType = window.createTextEditorDecoration
   }
 );
 
+const jdk8Jabba = "adopt@1.8";
+const jdk8Action = "JDK8";
+
+const jdk11Jabba = "adopt@1.11";
+const jdk11Action = "JDK11";
+
+const defaultJabbaVer = "0.11.2";
```
Comment (gabro) [inline src/extension.ts]: nit:

```suggestion
const defaultJabbaVersion = "0.11.2";
```

File: src/extension.ts [hunk 23]
```diff
@@ -87,16 +97,53 @@ export async function activate(context: ExtensionContext) {
...
+
+      const jabba_ver = defaultJabbaVer;
+      showJabbaMessage(message, jabba_ver);
+    });
+  commands.executeCommand("setContext", "metals:enabled", true);
+}
+
+function showJabbaMessage(
+  message: string,
+  jabba_ver: string = defaultJabbaVer
```
Comment (gabro) [inline src/extension.ts]: nit: 
```suggestion
  jabbaVersion: string = defaultJabbaVersion
```

File: src/extension.ts [hunk 24]
```diff
@@ -87,16 +97,53 @@ export async function activate(context: ExtensionContext) {
...
           commands.executeCommand("workbench.action.openSettings");
+          break;
         }
-      });
+        case jdk8Action: {
+          window.withProgress({
+            location: ProgressLocation.Notification,
+            title: `Installing JDK 8, please wait...`,
+            cancellable: true
+          }, (progress, _) => {
```
Comment (gabro) [inline src/extension.ts]: nit: `progress` is unused

```suggestion
          }, () => {
```

File: src/extension.ts [hunk 25]
```diff
@@ -87,16 +97,53 @@ export async function activate(context: ExtensionContext) {
...
+            return jdkinstaller.installJava(jdk8Jabba, jabba_ver).then(updateJavaConfig);
+          })
+          break;
+        }
+        case jdk11Action: {
+          window.withProgress({
+            location: ProgressLocation.Notification,
+            title: `Installing JDK 11, please wait...`,
+            cancellable: true
+          }, (progress, _) => {
```
Comment (gabro) [inline src/extension.ts]: nit: `progress` is unused

```suggestion
          }, () => {
```

File: src/extension.ts [hunk 26]
```diff
@@ -110,7 +157,15 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
...
-        `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
+      `To enable Metals, remove the file ${dottyArtifact} and run 'Reload window'`
+    );
+    return;
+  }
+
+  if (!javaHome || javaHome.trim().length == 0 || !fs.existsSync(javaHome)) {
+    showJabbaMessage(
+      "Incorrect 'Java Home' setting. To fix this problem, update the 'Java Home' setting to point to a Java 8 or Java 11 home directory " +
+      "or select a version to install automatically (window will need to be reloaded afterwards)."
```
Comment (gabro) [inline src/extension.ts]: nit: same comment as above

File: src/extension.ts [hunk 27]
```diff
@@ -228,6 +282,10 @@ function fetchAndLaunchMetals(context: ExtensionContext, javaHome: string) {
   );
 }
 
+function updateJavaConfig(javaHome: string, global: boolean = true) {
```
Comment (gabro) [inline src/extension.ts]: instead of a boolean, I would prefer this to use `ConfigurationTarget` instead of `boolean`.

File: src/jdkInstaller.ts [hunk 28]
```diff
@@ -0,0 +1,97 @@
+import * as path from "path";
+import * as request from "request";
+import * as fs from "fs";
+import * as mkdirp from "mkdirp";
+import * as pcp from "promisify-child-process";
+import { OutputChannel } from "vscode";
+
+export class JdkInstaller {
```
Comment (gabro) [inline src/jdkInstaller.ts]: I would prefer to refactor this not to be a class.

The only purpose of this class is to close over `out` which can simply be a parameter of `installJava` and then we can have a single function. All other class members can be simple values/functions inside this module.

I can work on this refactor @tgodizk

File: src/jdkInstaller.ts [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+
+  constructor(out: OutputChannel) {
+    this.out = out;
+  }
+
+  installJava(javaVersion: string, jabbaVersion: string) {
+    const jabbaUrl = `https://github.com/shyiko/jabba/releases/download/${jabbaVersion}/jabba-${jabbaVersion}-${this.jabbaUrlSuffix()}`;
+    const jabba = path.join(this.bin, this.jabbaName());
+    return new Promise<void>((res, rej) => {
+      mkdirp(this.bin, (err, made) => {
```
Comment (gabro) [inline src/jdkInstaller.ts]: minor, but we can probably use Node's `promisify` method here:


```
promisify(mkdirp)(this.bin)
```

I can work on this as well

**Comment by tgodzik** [comment]
> I find the code a bit non-idiomatic and I've left a few comments around. @tgodzik if you're with it I can rework some pieces a bit and directly commit on this PR, wdyt?

Sure, if you have time that would be awesome! I am still struggling a bit with typescript :sweat_smile:

File: src/extension.ts (hunk 20 above)
Comment (tgodzik) [reply src/extension.ts]: The user will be prompted - it's just to warn them about it

**Comment by gabro** [comment]
Superseded by #186
